config/v1/types_infrastructure.go

```diff
@@ -600,9 +600,300 @@ type OvirtPlatformStatus struct {
+
+const (
+	// HostGroupFailureDomain as a type allows the use of a group of ESXi hosts
+	// to be represented as a failure domain zone. When using this
+	// case it is expected that region would be a cluster.
+	// HostGroups within vCenter must be preconfigured and
+	// assigned in the topology.
+	HostGroupFailureDomain VSpherePlatformFailureDomainType = "HostGroup"
+
+	// ComputeClusterFailureDomain failure domain can either be a zone or region.
+	// The vCenter cluster is required to preconfigured and
+	// assigned in the topology.
+	ComputeClusterFailureDomain VSpherePlatformFailureDomainType = "ComputeCluster"
+
+	// DatacenterFailureDomain failure domain can be only be a region. The vcenter
+	// datacenter is required to be preconfigred and assigned
+	// in the topology. If used the zone would be of type ComputeCluster.
+	DatacenterFailureDomain VSpherePlatformFailureDomainType = "Datacenter"
+)
+
+// VSpherePlatformFailureDomainSpec holds the region and zone failure domain and
+// the vCenter topology of that failure domain.
+type VSpherePlatformFailureDomainSpec struct {
+	// name defines the name of the VSpherePlatformFailureDomainSpec
+	// This name is arbitrary but will be used
+	// in VSpherePlatformDeploymentZone for association.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=256
+	Name string `json:"name"`
+
+	// region defines a VSpherePlatformFailureDomain which
+	// includes the name of the vCenter tag, the failure domain type
+	// and the name of the vCenter tag category.
+	// +kubebuilder:validation:Required
+	Region VSpherePlatformFailureDomain `json:"region"`
+
+	// zone defines a VSpherePlatformFailureDomain which
+	// includes the name of the vCenter tag, the failure domain type
+	// and the name of the vCenter tag category.
+	// +kubebuilder:validation:Required
+	Zone VSpherePlatformFailureDomain `json:"zone"`
+
+	// Topology describes a given failure domain using vSphere constructs
+	// +kubebuilder:validation:Required
+	Topology VSpherePlatformTopology `json:"topology"`
+}
+
+// VSpherePlatformFailureDomain holds the name of the associated tag, the type
+// of the failure domain, and the vCenter tag category associated with this
+// failure domain.
+type VSpherePlatformFailureDomain struct {
+	// name is the name of the vCenter tag that represents this failure domain
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=80
+	Name string `json:"name"`
+
+	// type is the name of the failure domain type, which includes
+	// Datacenter, ComputeCluster and HostGroup
+	// +kubebuilder:validation:Required
+	Type VSpherePlatformFailureDomainType `json:"type"`
+
+	// tagCategory is the category used for the tag
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=80
+	TagCategory string `json:"tagCategory"`
```

Comment:
One of the concerns @jsafrane raised was - since we are not enforcing anything on these tag category names, what happens if customers names them inconsistently? Such as say:

1. `FailureDomain A` - tag is "regionA", category is" k8s-region", for zone - tag is "zoneA", category is "k8s-zone"
2. `FailureDomain B` - for region (tag is `regionA` and category is "k8s-region"), for zone (tag is "zoneB" and category is "locationZone")

In this case, categories of zones isn't matching and could create problems in determining topology of storage/nodes. I have not tested the behavior of CSI driver if categories are named inconsistently.   In a ideal case, for `failureDomain B` - a reasonable person would name category of the zone to be `k8s-zone` (basically same as category name in `failureDomainA`). 

So question is - should we enforce some kind of restriction on these names? Are we allowing any random string for a domain is too broad and will cause problems?

Reply:
Basically I think that - if user is using tag category - "k8s-region" for region, then they should use same tag category for all the regions in the cluster. Values can be different but category names should be the same. Allowing different "ways" of naming tag categories opens us up for various kind of corner cases. 

I was also thinking of picking a convention for these categories and sticking with it. So basically, you can't call your region category anything other than `k8s-region`, so on and so forth. What do you think?

Reply:
Updated removing the ability to set the tag category.
The `Region` and `Zone` are now just strings for the tag name.

Reply:
Is the category name now `openshift-zone` and `openshift-region` hardcoded values? Any other values are not accepted?

Reply:
Going to create a PR for the installer now to hardset the cloud-config labels to openshift-region, openshift-zone.
Will also update the enhancement

It won't be definable so not sure how other values would be accepted.

Reply:
https://github.com/openshift/installer/pull/6273

Reply:
Thank you. much neater. If need be we can relax this later.